Feat/dexalot candles feed

yancong001 opened the pull request:
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:

add dexalot candles feed

**Tests performed by the developer**:
unittest, candles_example.py test


**Tips for QA testing**:
![image](https://github.com/user-attachments/assets/b1c87165-72db-473a-b7fb-c5e23e7ef9fa)

The data returned by dexalot may be None, and it may also be None on the web page.

Modify candles_example.py's code for testing

```. 
    eth_1m_candles = CandlesFactory.get_candle(CandlesConfig(connector="dexalot", trading_pair="ETH-USDC", interval="5m", max_records=1000))
    eth_1h_candles = CandlesFactory.get_candle(CandlesConfig(connector="dexalot", trading_pair="ETH-USDC", interval="1h", max_records=1000))
    eth_1w_candles = CandlesFactory.get_candle(CandlesConfig(connector="dexalot", trading_pair="ETH-USDC", interval="1d", max_records=200))

```

nikspz commented:
commit 1451501

Wrong date on candles 

Steps:

1. Clone PR7394
2. start script below
[candles_exampled2.zip](https://github.com/user-attachments/files/18588374/candles_exampled2.zip)


Actual:
Review wrong timestamp (not 1h and 5m diffrence also)
![image](https://github.com/user-attachments/assets/337b8eb4-2fb8-4188-acef-f8f661753a41)

yancong001 commented:
After the big update of Candle in June, the timestamp was used in a uniform unit, so I also updated Candles_example.py to second. Now I have fixed the above problems. Please use the latest candles_example.py for testing.Or replace `candles_df["timestamp"] = pd.to_datetime(candles_df["timestamp"], unit="ms")
to candles_df["timestamp"] = pd.to_datetime(candles_df["timestamp"], unit="s")
` in your script candles_exampled2.zip

nikspz commented:
- Cloned and installed PR7394
- set up candles_example using dexalot
    - ETH-USDC 5m,15m,30m,1h,4h,1d Candles OHLC matched with exchange, reviewed for `latest candle close price may not match with exchange`
- download 5m,15m,1h, 1d dexalot candles using download_script
- Checked AVAX-USDC 
  - `5m` candles, got `NaN for candles` have values in the exchange( not matched)
  - `15m` candles not showing live updates (`latest candle not showing`)
  - `30m` candles - close price not matched with exchange
  
scripts:
[7394candlesdexalot.zip](https://github.com/user-attachments/files/18676628/7394candlesdexalot.zip)

  
- running v2 with dexalot candles:
  - got an issue with order price (using dexalot candles and dexalot rate oracle

[logs_dex1.log](https://github.com/user-attachments/files/18676618/logs_dex1.log)



AVAX-USDC:
![image](https://github.com/user-attachments/assets/4ac76979-e22b-4a5b-98a4-67b8f9c2693d)
![image](https://github.com/user-attachments/assets/f337b4da-e281-4a07-bc90-1e2719a1cb19)
![image](https://github.com/user-attachments/assets/d6535d2f-be40-4f1b-8856-005454b52a3e)

dexalot candles on v2 (pmm_dynamic)

[logs_dex1.log](https://github.com/user-attachments/files/18676606/logs_dex1.log)

![image](https://github.com/user-attachments/assets/2bb7105d-e694-4cc3-b3d1-fb2e14895c1f)

yancong001 commented:
![3961739068656_ pic](https://github.com/user-attachments/assets/0edb91c2-18a7-47ac-89b4-2017ed4cd7e6)

I tested 5 minutes of data and the results showed that there was a considerable delay in the push of websockets, and not every piece of data was pushed in real time.
As shown in Fig., 1. The transaction time of "close:24.638" on the webpage is 10:35:12, and the data time of websocket pushing kline is 10:36:03, a difference of 49 seconds.
2. The latest transaction data of 24.615 was not pushed in time.

>  got an issue with order price (using dexalot candles and dexalot rate oracle
The reason for this issue could be the presence of Zero data in the candle

nikspz commented:
> ![3961739068656_ pic](https://private-user-images.githubusercontent.com/29802081/411281021-0edb91c2-18a7-47ac-89b4-2017ed4cd7e6.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkzMDE1OTksIm5iZiI6MTczOTMwMTI5OSwicGF0aCI6Ii8yOTgwMjA4MS80MTEyODEwMjEtMGVkYjkxYzItMThhNy00N2FjLTg5YjQtMjAxN2VkNGNkN2U2LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjExVDE5MTQ1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMyMmNlMGQxNGE1NmRhOTUwNjg3M2Y1M2VhN2U5YWI2OWQwNTkyODE5ZTQzMDgyYTJkMGM4ZDZiM2Q4YmYwNjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MgTJlW9XxuL2e_rtBCAg014zOxM4nZxuySHVQwjMyZo)
> 
> I tested 5 minutes of data and the results showed that there was a considerable delay in the push of websockets, and not every piece of data was pushed in real time. As shown in Fig., 1. The transaction time of "close:24.638" on the webpage is 10:35:12, and the data time of websocket pushing kline is 10:36:03, a difference of 49 seconds. 2. The latest transaction data of 24.615 was not pushed in time.
> 
> > got an issue with order price (using dexalot candles and dexalot rate oracle
> > The reason for this issue could be the presence of Zero data in the candle

Yes, there's delay around 1 minute at least

commit [9abcfb5](https://github.com/hummingbot/hummingbot/pull/7394/commits/9abcfb5e96f8cbc3a974cd36142ac427ff5b49e5)


![dexalot delay 5m delay](https://github.com/user-attachments/assets/9e06234c-fbaf-4dd0-869b-7bbf205f4176)
![dexalot delay 1 min](https://github.com/user-attachments/assets/7b5aa9a3-c6cf-43a9-854a-22725e4a4778)

yancong001 commented:
The problem of delay does not seem to be officially planned to be solved recently, so ws needs to evaluate the feasibility of Dexalot Candle.

yancong001 commented:
i will continue

yancong001 commented:
I have updated and completed it, and there are no problems in the actual test. However, there seem to be many errors in the unittest. Is the team updating the relevant unittest of candle feed? Do I need to solve my unittest errors after the team completes the update?

nikspz commented:
> I have updated and completed it, and there are no problems in the actual test. However, there seem to be many errors in the unittest. Is the team updating the relevant unittest of candle feed? Do I need to solve my unittest errors after the team completes the update?

We migrated to pytest here: https://github.com/hummingbot/hummingbot/pull/7403
hi @yancong001 please fix test failing: https://github.com/hummingbot/hummingbot/actions/runs/14476139941/job/40602156085?pr=7394

yancong001 commented:
fixed

nikspz reviewed (approved):
- commit e2535c4
    - Test performed:
        - set up candles_example script using dexalot candles
            - 5m candles matched with exchange
            - 15m candles matched with exchange
            - 30m candles matched with exchange
            - 1h candles matched with exchange
            - 4h candles matched with exchange
            - 1d candles matched with exchange
        - able to run v2 strategy using dexalot candles
        - manually built docker image successfully

Note:
Noticed 5m (and sometimes 15m) candles could be empty on exchange side:
![image](https://github.com/user-attachments/assets/29a818cf-fe95-4c3b-87cf-74da968f1507)
![image](https://github.com/user-attachments/assets/c317dc42-9b39-4a0d-a231-63d4ee11bd33)

cardosofede reviewed (approved):
Thanks @yancong001
